messaging/messaging_batch.go

```diff
@@ -0,0 +1,273 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package messaging
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"mime"
+	"mime/multipart"
+	"net/http"
+	"net/textproto"
+
+	"firebase.google.com/go/internal"
+)
+
+const multipartBoundary = "__END_OF_PART__"
+
+// SendResponse represents the status of an individual message that was sent as part of a batch
+// request.
+type SendResponse struct {
+	Success   bool
+	MessageID string
+	Error     error
+}
+
+// BatchResponse represents the response from the `SendAll()` and `SendMulticast()` APIs.
+type BatchResponse struct {
+	SuccessCount int
+	FailureCount int
+	Responses    []*SendResponse
+}
+
+// SendAll sends all the messages in the given array via Firebase Cloud Messaging.
+//
+// The messages array may contain up to 100 messages. SendAll employs batching to send the entire
+// array of mssages as a single RPC call. Compared to the `Send()` function,
+// this is a significantly more efficient way to send multiple messages. The responses
+// list obtained from the return value corresponds to the order of input messages. An error from
+// SendAll indicates a total failure -- i.e. none of the messages in the array could be sent.
+// Partial failures are indicated by a `BatchResponse` return value.
+func (c *Client) SendAll(ctx context.Context, messages []*Message) (*BatchResponse, error) {
+	return c.sendBatch(ctx, messages, false)
+}
+
+func (c *Client) sendBatch(
+	ctx context.Context, messages []*Message, dryRun bool) (*BatchResponse, error) {
+
+	if len(messages) == 0 {
+		return nil, errors.New("messages must not be nil or empty")
+	}
+
+	if len(messages) > 100 {
+		return nil, errors.New("messages must not contain more than 100 elements")
+	}
+
+	request, err := c.newBatchRequest(messages, dryRun)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.client.Do(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.Status != http.StatusOK {
+		return nil, handleFCMError(resp)
+	}
+
+	return newBatchResponse(resp)
+}
+
+// part represents a HTTP request that can be sent embedded in a multipart batch request.
+//
+// See https://cloud.google.com/compute/docs/api/how-tos/batch for details on how GCP APIs support multipart batch
+// requests.
+type part struct {
+	method  string
+	url     string
+	headers map[string]string
+	body    interface{}
+}
+
+// multipartEntity represents an HTTP entity that consists of multiple HTTP requests (parts).
+type multipartEntity struct {
+	parts []*part
+}
+
+func (c *Client) newBatchRequest(messages []*Message, dryRun bool) (*internal.Request, error) {
+	url := fmt.Sprintf("%s/projects/%s/messages:send", c.fcmEndpoint, c.project)
+	headers := map[string]string{
+		"X-GOOG-API-FORMAT-VERSION": "2",
```

Comment:
Should this be in a constant somewhere? It seems like the kind of header multiple APIs would use.

Reply:
Done